src/plugins/dashboard/server/saved_objects/dashboard_migrations.ts

```diff
@@ -134,26 +147,91 @@ function createExtractPanelReferencesMigration(
+  // object has panelsJSON in all previous versions of kibana.
+  if (typeof attributes.panelsJSON !== 'string') {
+    return attributes;
+  }
+  const panels = JSON.parse(attributes.panelsJSON) as SavedDashboardPanel[];
+  // Same here, prevent failing saved object import if ever panels aren't an array.
+  if (!Array.isArray(panels)) {
+    return attributes;
+  }
+  const newPanels: SavedDashboardPanel[] = [];
+  panels.forEach((panel) => {
+    // Convert each panel into a state that can be passed to EmbeddablesSetup.migrate
+    const originalPanelState = convertSavedDashboardPanelToPanelState<ValueOrReferenceInput>(panel);
+    if (originalPanelState.explicitInput.attributes) {
+      // If this panel is by value, migrate the state using embeddable migrations
+      const migratedInput = deps.embeddable.migrate(
+        {
+          ...originalPanelState.explicitInput,
+          type: originalPanelState.type,
+        },
+        version
+      );
+      // Convert the embeddable state back into the panel shape
+      newPanels.push(
+        convertPanelStateToSavedDashboardPanel(
+          {
+            ...originalPanelState,
+            explicitInput: { ...migratedInput, id: migratedInput.id as string },
+          },
+          version
+        )
+      );
+    } else {
+      newPanels.push(panel);
+    }
+  });
+  return {
+    ...doc,
+    attributes: {
+      ...attributes,
+      panelsJSON: JSON.stringify(newPanels),
+    },
+  };
+};
+
 export interface DashboardSavedObjectTypeMigrationsDeps {
   embeddable: EmbeddableSetup;
 }
 
 export const createDashboardSavedObjectTypeMigrations = (
   deps: DashboardSavedObjectTypeMigrationsDeps
-) => ({
-  /**
-   * We need to have this migration twice, once with a version prior to 7.0.0 once with a version
-   * after it. The reason for that is, that this migration has been introduced once 7.0.0 was already
-   * released. Thus a user who already had 7.0.0 installed already got the 7.0.0 migrations below running,
-   * so we need a version higher than that. But this fix was backported to the 6.7 release, meaning if we
-   * would only have the 7.0.1 migration in here a user on the 6.7 release will migrate their saved objects
-   * to the 7.0.1 state, and thus when updating their Kibana to 7.0, will never run the 7.0.0 migrations introduced
-   * in that version. So we apply this twice, once with 6.7.2 and once with 7.0.1 while the backport to 6.7
-   * only contained the 6.7.2 migration and not the 7.0.1 migration.
-   */
-  '6.7.2': flow(migrateMatchAllQuery),
-  '7.0.0': flow(migrations700),
-  '7.3.0': flow(migrations730),
-  '7.9.3': flow(migrateMatchAllQuery),
-  '7.11.0': flow(createExtractPanelReferencesMigration(deps)),
-});
+): SavedObjectMigrationMap => {
+  const embeddableMigrations = deps.embeddable
+    .getMigrationVersions()
+    .filter((version) => semver.gt(version, '7.12.0'))
```

Comment:
Do we even need to filter here, since there aren't any embeddable migrations before 7.12?

Reply:
Hum, that is a good point. It could potentially work either way, but maybe it's worth leaving it in just in case an embeddable migration gets added to an older version?

Reply:
we are not maintaining older versions so i don't think we'll do that